### PR TITLE
Add support for ADT 3.0 and ktlin tasks for multidimension projects

### PR DESCRIPTION
**Pull request by originx**
Add support for Android build tools 3.0 and ktlin tasks for multidimension builds.

new google repo is added since this is now a maven repo where google
hosts all of its tools from now on.

related to:
#29

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -13,12 +13,13 @@ version = "2.2.2-SNAPSHOT"
...
 repositories {
     jcenter()
+    google()
 }
 
 dependencies {
     compileOnly(gradleApi())
     compileOnly(kotlin("gradle-plugin", "1.1.1"))
-    compileOnly("com.android.tools.build:gradle:2.3.3")
+    compileOnly("com.android.tools.build:gradle:3.0.0-rc2")
```
Comment (Tapchicoma) [inline build.gradle.kts]: `3.0.0` is already available, please update to it.

File: src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt [hunk 2]
```diff
@@ -63,28 +64,38 @@ open class KtlintPlugin : Plugin<Project> {
...
                 val ktLintConfig = createConfiguration(target, extension)
-                
-                target.extensions.findByType(BaseExtension::class.java)?.sourceSets?.forEach {
-                    val kotlinSourceDir = it.java.sourceFiles
-                    val runArgs = it.java.srcDirs.map { "${it.path}/**/*.kt" }.toMutableSet()
+
+                val variantManager = when {
+                    target.plugins.hasPlugin(AppPlugin::class.java) -> target.plugins.findPlugin(AppPlugin::class.java)?.variantManager
+                    target.plugins.hasPlugin(LibraryPlugin::class.java) -> target.plugins.findPlugin(LibraryPlugin::class.java)?.variantManager
+                    target.plugins.hasPlugin(InstantAppPlugin::class.java) -> target.plugins.findPlugin(LibraryPlugin::class.java)?.variantManager
```
Comment (leinardi) [inline src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt]: Hi @originx, is it wanted to use always the LibraryPlugin? Probably it will be better to use the specific Plugin, in this case InstantAppPlugin.

Comment (originx) [reply src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt]: yep thanks, rush of duplicating lines, fixed now :)

**Comment by Tapchicoma** [comment]
@originx can you also update CHANGELOG.md?

**Comment by leinardi** [comment]
@Tapchicoma what about kotlin version? Any reason why the project is using 1.1.1 instead of 1.1.51?

File: build.gradle.kts (hunk 1 above)
Comment (originx) [reply build.gradle.kts]: sure

**Comment by Tapchicoma** [comment]
@leinardi I have the same question, can't answer it by now. But it is not the scope of this PR.

**Comment by originx** [comment]
review comments done, please recheck,

cheers :)

**Comment by Tapchicoma** [comment]
I want to test this changes later, when I will at home. Overall looks good.

File: src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 package org.jlleitschuh.gradle.ktlint
 
-import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.*
```
Comment (JLLeitschuh) [inline src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt]: Don't use star imports please

Comment (Tapchicoma) [reply src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt]: Ktlint-gradle should run ktlint on itself 😄

**Comment by Tapchicoma** [comment]
Checked, works ok.

@JLLeitschuh this PR breaks support for older android plugin (2.3.3). I propose release current master with fix for newer ktlint versions and then make a new release with this PR.

**Comment by JLLeitschuh** [comment]
This fix is going to require a major version bump of the plugin because one of our dependencies has incremented its major version number.

**Comment by JLLeitschuh** [comment]
This now has merge conflicts (sorry). We may want to merge #36 first.

**Comment by JLLeitschuh** [comment]
I just made your merge conflicts much worse. Sorry.

File: src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt (hunk 3 above)
Comment (Tapchicoma) [reply src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt]: Fixed in 7ddc8b03b562446d64edebf1b4d7eec3f0bdbb09

File: CHANGELOG.md [hunk 4]
```diff
@@ -19,7 +20,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ### Changed
  - Define a different output file for each sourceSet
 ### Fixed
- - Fixed plugin doesn't apply custom reporter for ktlint versions >0.10.x (#28)
+- Fixed plugin doesn't apply custom reporter for ktlint versions >0.10.x (#28)
```
Comment (JLLeitschuh) [inline CHANGELOG.md]: Unnecessary whitespace change?

Comment (Tapchicoma) [reply CHANGELOG.md]: Oh, damn, will update

File: plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt [hunk 5]
```diff
@@ -63,28 +68,37 @@ open class KtlintPlugin : Plugin<Project> {
...
-                target.extensions.findByType(BaseExtension::class.java)?.sourceSets?.forEach {
-                    val kotlinSourceDir = it.java.sourceFiles
-                    val runArgs = it.java.srcDirs.map { "${it.path}/**/*.kt" }.toMutableSet()
+                val variantManager = when {
+                    target.plugins.hasPlugin(AppPlugin::class.java) -> target.plugins.findPlugin(AppPlugin::class.java)?.variantManager
+                    target.plugins.hasPlugin(LibraryPlugin::class.java) -> target.plugins.findPlugin(LibraryPlugin::class.java)?.variantManager
+                    target.plugins.hasPlugin(InstantAppPlugin::class.java) -> target.plugins.findPlugin(InstantAppPlugin::class.java)?.variantManager
+                    target.plugins.hasPlugin(FeaturePlugin::class.java) -> target.plugins.findPlugin(FeaturePlugin::class.java)?.variantManager
+                    target.plugins.hasPlugin(TestPlugin::class.java) -> target.plugins.findPlugin(TestPlugin::class.java)?.variantManager
+                    else -> throw StopExecutionException("Must be applied with 'android' or 'android-library' plugin.")
```
Comment (JLLeitschuh) [inline plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt]: Are all of these types guaranteed to be on the classpath when either the `android` or the `android-library` plugin is applied?

If only the `android` plugin is applied will you get a `ClassDefNotFoundException` for the `AppPlugin` class?

File: CHANGELOG.md (hunk 4 above)
Comment (Tapchicoma) [reply CHANGELOG.md]: Check this commit d26f377041144ac55b188d5dcc475da472ebb1a8

File: plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt (hunk 5 above)
Comment (Tapchicoma) [reply plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt]: Yes, everything is ok. I've tested on my local test project.

**Comment by Tapchicoma** [comment]
No, definetely I will add android project sample but in a separate PR.
